components/server/src/express/ws-ping-pong-handler.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+ import * as express from 'express';
+ import * as websocket from 'ws';
+import { Disposable, DisposableCollection } from "@gitpod/gitpod-protocol";
+import { repeat } from "@gitpod/gitpod-protocol/lib/util/repeat";
+import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
+import { WsNextFunction, WsRequestHandler } from './ws-handler';
+
+/**
+ * This class provides a websocket handler that manages ping-pong behavior for all incoming websocket requests.
+ * Clients that to not respond in time are terminated.
+ */
+export class WsPingPongHandler implements Disposable {
+
+    protected readonly disposables: DisposableCollection = new DisposableCollection();
+    protected readonly clients: Set<websocket> = new Set();
+
+    start(): void {
+        // implement heartbeating closely following https://www.npmjs.com/package/ws#how-to-detect-and-close-broken-connections
+        const INTERVAL = 30000;
+        const TIMEOUT = INTERVAL;
+        const CLOSING_TIMEOUT = INTERVAL;
+        const timer = repeat(async () => {
+            this.clients.forEach((ws) => {
+                try {
+                    switch (ws.readyState) {
+                        case websocket.CLOSED:
+                            // ws should not be in the clients list anymore
+                            log.warn("websocket in strange state", { readyState: ws.readyState });
```

Comment:
In the "strange state" cases the `ws` stays in the list of `clients`. The only way it's ever removed is in the `onClose` handler below. Should/can we do something to avoid a memory leak when for some reason `onClose` is never fired?

Reply:
IMO the likelihood is rather small, as that would mean that we already have this in prod.
Nonetheless I added this branch and `warn` message to detect exactly this case, which we could not before. If we see it being logged we can add (delayed) termination here.

Note that we have delayed termination for `CLOSING` now, which is more relevant because `ws` holds an internal timer.